### ISSUE TITLE
Can I use this library together with Xamarin.Firebase.Messaging?

### ISSUE THREAD

**Issue by suariana** [AdamEssenmacher/GoogleApisForiOSComponents#58]
I have problem when building my project when i use Adame.Firebas.IOS.CloudMessaging
i've followed step of longpath issues but i still cannot build my project
i use Xamarin.Firebase.Messaging for android and Adame.Firebas.IOS.CloudMessaging for iOS
can both of library work together in the same project?
thank you

**Comment by AdamEssenmacher** [comment]
No. AdamE.iOS.Firebase.Core is incompatible with Xamarin.iOS.Firebase.Core, and also with any other package that includes common Firebase/Google iOS dependencies such as GTMSessionFetcher. This is documented in the project readme.

**Comment by suariana** [comment]
![image](https://github.com/user-attachments/assets/c9b2f435-3813-41ac-a4c7-532e391c96d0)
![image](https://github.com/user-attachments/assets/ca040180-28b0-4b43-b4f6-977c36248742)
![image](https://github.com/user-attachments/assets/c6addf7f-c7e9-40bb-826f-845ca239e8fb)
may i know which one of that package that cannot work together with Plugin.Firebase?
because i still cannot build my project
im using windows 10
thank you

**Comment by suariana** [comment]
![image](https://github.com/user-attachments/assets/282352df-f0f1-44b6-87dc-ffcc5b2c134a)
![image](https://github.com/user-attachments/assets/59fc4c00-9b14-4302-9a11-288c4dd55ca8)
![image](https://github.com/user-attachments/assets/8b835c8d-dd2d-4454-bd22-bcc85151a210)
may i know which one of that package is not compatible with Plugin.Firebase?
because i still cannot build my project, there's no error long path or anythink on error list
thank you

**Comment by maonaoda** [comment]
@AdamEssenmacher 
Does this mean that it cannot be used together with Xamarin.Google.iOS.SignIn?

**Comment by AdamEssenmacher** [comment]
@suariana I think I misunderstood your first question, so let me try again.

The nugets published from this repository are binding libraries for native iOS SDKs--they have nothing to do with Android. They only make sense in the context of iOS. They shouldn't be included in any builds that target any other platform. If you're referencing these packages directly in a multi-target project (e.g. MAUI), then the packages should be conditionally included (see the readme).

Based on your screenshot, the build is hanging while XamarinBuildDownload is trying to extract GAppM. Check the project readme for troubleshooting steps around Windows and 7Zip.

**Comment by AdamEssenmacher** [comment]
@maonaoda Mixing Xamarin.Google.iOS.SignIn with any AdamE.Firebase.iOS.* NuGet package is not recommended. This is because Xamarin.Google.iOS.SignIn depends on Xamarin.Firebase.iOS.Core, while AdamE.Firebase.iOS.Core is a replacement for Xamarin.Firebase.iOS.Core. Using both in the same project will cause both to link to the same set of native libraries (e.g., GTMSessionFetcher and PromisesObjC) but with different versions. This could lead to build-time issues or, worse, runtime conflicts. For more details about these incompatibilities, please refer to the project README.

If you're interested in adding Google Sign-In support to this project, I'd gladly review a community PR. Google Sign-In would likely follow a similar implementation to Google Maps, which was recently added. Reviewing the commit history for that feature could serve as a helpful guide.

Alternatively, if you'd like to see this feature prioritized, I offer one-time sponsorship options on my GitHub profile that could cover an afternoon of my time.

**Comment by maonaoda** [comment]
`Based on your screenshot, the build is hanging while XamarinBuildDownload is trying to extract GAppM. Check the project readme for troubleshooting steps around Windows and 7Zip.`
same with 
https://github.com/dotnet/android-libraries/issues/1047

**Comment by maonaoda** [comment]
> @maonaoda Mixing Xamarin.Google.iOS.SignIn with any AdamE.Firebase.iOS.* NuGet package is not recommended. This is because Xamarin.Google.iOS.SignIn depends on Xamarin.Firebase.iOS.Core, while AdamE.Firebase.iOS.Core is a replacement for Xamarin.Firebase.iOS.Core. Using both in the same project will cause both to link to the same set of native libraries (e.g., GTMSessionFetcher and PromisesObjC) but with different versions. This could lead to build-time issues or, worse, runtime conflicts. For more details about these incompatibilities, please refer to the project README.
> 
> If you're interested in adding Google Sign-In support to this project, I'd gladly review a community PR. Google Sign-In would likely follow a similar implementation to Google Maps, which was recently added. Reviewing the commit history for that feature could serve as a helpful guide.
> 
> Alternatively, if you'd like to see this feature prioritized, I offer one-time sponsorship options on my GitHub profile that could cover an afternoon of my time.

Thank you for your detailed explanation, 
I was able to get the build to work by replacing the GoogleAppMeasurement files,
but do have some concerns about the runtime issue （may be）

**Comment by suariana** [comment]
> @suariana I think I misunderstood your first question, so let me try again.
> 
> The nugets published from this repository are binding libraries for native iOS SDKs--they have nothing to do with Android. They only make sense in the context of iOS. They shouldn't be included in any builds that target any other platform. If you're referencing these packages directly in a multi-target project (e.g. MAUI), then the packages should be conditionally included (see the readme).
> 
> Based on your screenshot, the build is hanging while XamarinBuildDownload is trying to extract GAppM. Check the project readme for troubleshooting steps around Windows and 7Zip.

Thank you, my project can be built now
but, is this library compatible with dotnet 8?
because i cannot receive notification on ios?
i only using adame firebase cloud messaging for my project to replace old Xamarin.Firebase.IOS.CloudMessaging

![image](https://github.com/user-attachments/assets/017e6317-3940-4b93-b85a-d7d84a0d3bb2)
![image](https://github.com/user-attachments/assets/48c7ac98-8802-41da-a49b-44f47a86e056)

**Comment by AdamEssenmacher** [comment]
It's important to recognize that the AdamE version of FCM advances the underlying native SDK version two major versions (from 8.10 to 10.16+). Native configuration requirements changed.

I'm guessing your problem stems from config changes related to method swizzling.

Yes, these bindings work on .net 6 through 9.

**Comment by suariana** [comment]
using Firebase.CloudMessaging;
using Foundation;
using UIKit;
using UserNotifications;

namespace FirebaseNotification.Platforms.iOS
{

    public class FirebaseService : NSObject, IMessagingDelegate
    {

        internal static void Initialize()
        {

            var fs = new FirebaseService();
            fs.Configure();

        }

        private void Configure()
        {
            Firebase.Core.App.Configure();
            if (UIDevice.CurrentDevice.CheckSystemVersion(10, 0))
            {
                var authOption = UNAuthorizationOptions.Alert | UNAuthorizationOptions.Badge | UNAuthorizationOptions.Sound;

                UNUserNotificationCenter.Current.RequestAuthorization(authOption, (granted, error) =>
                {
                    Console.WriteLine($"Registration is success: {granted}");

                    if (error != null)
                    {
                        Console.WriteLine($"Registration for push notifications error: {error.DebugDescription}");
                    }

                    if (granted && error == null)
                    {
                        this.InvokeOnMainThread(() =>
                        {
                            UIApplication.SharedApplication.RegisterForRemoteNotifications();
                            this.RegisterNotification();
                        });
                    }
                });


            }


        }

        private void RegisterNotification()
        {
            UNUserNotificationCenter.Current.Delegate = new Platforms.iOS.UserNotificationCenterDelegate();
            Messaging.SharedInstance.AutoInitEnabled = true;
            Messaging.SharedInstance.Delegate = this;
            GetToken();
        }

        private void GetToken()
        {
            var token = Messaging.SharedInstance.FcmToken;
            ImkFirebaseNotification.Current.InvokeTokenRefresh(token);
        }

        internal static void SubscribeToTopic(string topic)
        {
            Messaging.SharedInstance.Subscribe(topic);
        }

        internal static void UnsubscribeTopic(string topic)
        {
            Messaging.SharedInstance.Unsubscribe(topic);
        }

        [Export("messaging:didReceiveRegistrationToken:")]
        private void DidReceiveRegistrationToken(Messaging message, string regToken)
        {
            ImkFirebaseNotification.Current.InvokeTokenRefresh(regToken);
        }


    }
}





using Foundation;
using UserNotifications;

namespace FirebaseNotification.Platforms.iOS
{
    public class UserNotificationCenterDelegate : UNUserNotificationCenterDelegate
    {

        public static UNNotificationPresentationOptions CurrentNotificationPresentationOption { get; set; } = UNNotificationPresentationOptions.None;

        public override void DidReceiveNotificationResponse(UNUserNotificationCenter center, UNNotificationResponse response, Action completionHandler)
        {
            var userInfo = response.Notification.Request.Content.UserInfo;
            var parameters = GetParameters(userInfo);
            Current.InvokeNotificationOpened(parameters);
            completionHandler();
        }

        public override void WillPresentNotification(UNUserNotificationCenter center, UNNotification notification, Action<UNNotificationPresentationOptions> completionHandler)
        {
            var userInfo = notification.Request.Content.UserInfo;

            var parameters = GetParameters(userInfo);
            Current.InvokeNotificationReceived(parameters);

            //Tambil notif terus
            //completionHandler(UNNotificationPresentationOptions.Alert);

            // Tampil notif hanya yang penting
            if ((parameters.TryGetValue("priority", out var priority) && ($"{priority}".ToLower() == "high" || $"{priority}".ToLower() == "max")))
            {
                if (!CurrentNotificationPresentationOption.HasFlag(UNNotificationPresentationOptions.Alert))
                {
                    CurrentNotificationPresentationOption |= UNNotificationPresentationOptions.Alert;

                }
            }
            else if ($"{priority}".ToLower() == "default" || $"{priority}".ToLower() == "low" || $"{priority}".ToLower() == "min")
            {
                if (CurrentNotificationPresentationOption.HasFlag(UNNotificationPresentationOptions.Alert))
                {
                    CurrentNotificationPresentationOption &= ~UNNotificationPresentationOptions.Alert;

                }
            }
            completionHandler(CurrentNotificationPresentationOption);
        }


        private static IDictionary<string, string> GetParameters(NSDictionary data)
        {
            var parameters = new Dictionary<string, string>();

            var keyAps = new NSString("aps");
            var keyAlert = new NSString("alert");

            foreach (var val in data)
            {
                if (val.Key.Equals(keyAps))
                {
                    var aps = data.ValueForKey(keyAps) as NSDictionary;

                    if (aps != null)
                    {
                        foreach (var apsVal in aps)
                        {
                            if (apsVal.Value is NSDictionary)
                            {
                                if (apsVal.Key.Equals(keyAlert))
                                {
                                    foreach (var alertVal in apsVal.Value as NSDictionary)
                                    {
                                        parameters.Add($"aps.alert.{alertVal.Key}", $"{alertVal.Value}");
                                    }
                                }
                            }
                            else
                            {
                                parameters.Add($"aps.{apsVal.Key}", $"{apsVal.Value}");
                            }

                        }
                    }
                }
                else
                {
                    parameters.Add($"{val.Key}", $"{val.Value}");
                }

            }


            return parameters;
        }

    }
}


im sorry may i know that im using swizzling method or not?

**Comment by AdamEssenmacher** [comment]
Closing this now because the original issue has been solved and the conversation is starting to sprawl.

@suariana the FCM bindings work. If you're not getting notifications, it is almost certainly because your project is just misconfigured somehow. There are a few issues discussing the same sort of problem over at Plugin.Firebase that might push you in the right direction:

https://github.com/TobiasBuchholz/Plugin.Firebase/issues/324
https://github.com/TobiasBuchholz/Plugin.Firebase/issues/192
https://firebase.google.com/docs/cloud-messaging/ios/client#method_swizzling_in

**Comment by suariana** [comment]
@AdamEssenmacher  im sorry may i ask once again, when im using old Xamarin.Firebase.iOS.CloudMessaging i always received diferrence token when i uninstall and install the application
but when i changed the library to AdamE.Firebase.iOS.CloudMessaging i always received last token that generate from old Xamarin.Firebase.iOS.CloudMessaging 
the code that i used to get the token is like below

```
 private void GetToken()
 {
     var token = Messaging.SharedInstance.FcmToken;
     ImkFirebaseNotification.Current.InvokeTokenRefresh(token);
 }
```

is the code that used to get the token still same with old Xamarin.Firebase.iOS.CloudMessaging?
thankyou, because i still stuck with new library :(
